### PR TITLE
Fix #5450: Handle `this` and self reference in IDE

### LINKED ISSUE #5450: IDE: `this` and `self` dot not behave properly
Given the following source:

```scala
class A { self =>
  def fizz = self
  def buzz = this
}
```

### `textDocument/documentHighlight`
| Where's the mouse  | What's highlighted                   | What should be highlighted | ✅ / ❌ |
| -------------------- | ----------------------------- | ---------------------------- |:--------:|
| `A:L1`                        | `A:L1`, `self:L2`                       | `A:L1`                                      | ❌         |
| `self:L1`                    | `self:L1`                                     | `self:L1`, `self:L2`, `this:L3`| ❌         |
| `self:L2`                   | `A:L1`, `self:L2`                        | `self:L1`, `self:L2`, `this:L3`| ❌         |
| `this:L3`                   | `A:L1`, `self:L2`                        | `self:L1`, `self:L2`, `this:L3`| ❌         |

It appears that `self:L1` has a symbol of its own. Usages of `self` have a different symbol, though.

We do not handle this at all, so it looks like most interactions with the IDE that involve `this` or an alias to `this` give wrong results.

### PR DESCRIPTION
**Pull request by Duhemm**
> In the IDE, all the class definition now have a self definition. We use
> the symbol of this definition for `this` and references to the self
> definition.

The basic facts that make handling `this` and `self` hard in the IDE is this (it seems no one references the symbol of the self definition):

> Currently, a self definition has a symbol. The symbol of references to
> the self definition is the enclosing class symbol. The symbol of `this`
> is the enclosing class symbol.

I don't know what are the reasons for having it work this way, nor how hard it would be to have a scheme that would be easier to handle for the IDE.

I also tried to use the kind of the name to know what to select (grouping `this` and `self` by knowing that they have the same symbol and the names they carry have the same kind), but I ran into problems with imports that always have term names, IIRC.

Note that if this PR is merged, we will start to be able to rename `this`. I added support for `textDocument/prepareRename` (and added checks in `rename`), and will submit a PR with these changes this week.

### CODE REVIEW

**Comment by odersky** [comment]
> > Currently, a self definition has a symbol. The symbol of references to
the self definition is the enclosing class symbol. The symbol of this
is the enclosing class symbol.

> I don't know what are the reasons for having it work this way, nor how hard it would be to have a scheme that would be easier to handle for the IDE.

It's because the type of the self reference and of `this` is in each case a `ThisType`, which has a class as symbol. I believe this would be very hard to change.

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala [hunk 1]
```diff
@@ -282,6 +282,14 @@ trait TreeInfo[T >: Untyped <: Type] { self: Trees.Instance[T] =>
...
     case _ => p(tree)
   }
+
+  /** Is `tree` a reference to `this` or a self type? */
+  def isThisOrSelfReference(tree: Tree)(implicit ctx: Context): Boolean = tree match {
+    case This(_) => true
+    case Ident(nme) => nme.isTermName && tree.symbol.isType
+    case _ => false
+  }
+
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/ast/TreeInfo.scala]: Use the `isSelfSym` method instead. That's a more robust test.

**Comment by b-studios** [comment]
@Duhemm Hey, do you intend to work on this PR or should we close it?

**Comment by Duhemm** [comment]
I won't have the time to work on this unfortunately, so let's close.

**Pull request closed without merging**